query/src/sql/planner/semantic/type_check.rs

```diff
@@ -650,15 +651,20 @@ impl<'a> TypeChecker<'a> {
                 let op = ComparisonOp::try_from(op)?;
                 let (left, _) = self.resolve(left, None).await?;
                 let (right, _) = self.resolve(right, None).await?;
-
+                let mut data_type = BooleanType::new_impl();
+                if left.data_type() == DataTypeImpl::Null(NullType {})
+                    || right.data_type() == DataTypeImpl::Null(NullType {})
+                {
+                    data_type = NullType::new_impl();
+                }
                 Ok((
                     ComparisonExpr {
                         op,
                         left: Box::new(left),
                         right: Box::new(right),
                     }
                     .into(),
-                    BooleanType::new_impl(),
+                    data_type,
```

Comment:
I took a general look and this return value should not be needed. So https://github.com/datafuselabs/databend/pull/6043/files#diff-37b754ad52b168f02e969eb2486df14bde8efb8c819a218053a1dcf79bff6b50R654-R659 will not be needed

Reply:
Not exactly, this is part of the original type check algorithm. But since we decided to refactor the whole type system, so we didn't finish this.

It's better to keep that for now.